polly/src/event_manager.rs

```diff
@@ -0,0 +1,786 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+use epoll::Events;
+use pollable::{EventRegistrationData, EventSet, Pollable, PollableOp, PollableOpBuilder};
+use std::collections::HashMap;
+use std::fmt::Formatter;
+use std::io;
+use std::ops::{Deref, DerefMut};
+use std::os::unix::io::{AsRawFd, FromRawFd, RawFd};
+use std::sync::mpsc::{channel, Receiver, Sender, TryRecvError};
+use std::sync::{Arc, Mutex};
+use sys_util::EventFd;
+
+const EVENT_BUFFER_SIZE: usize = 128;
+const DEFAULT_EPOLL_TIMEOUT: i32 = 250;
+
+pub type Result<T> = std::result::Result<T, Error>;
+pub type WrappedHandler = Arc<Mutex<dyn EventHandler>>;
+
+pub enum Error {
+    /// Cannot create epoll fd.
+    EpollCreate,
+    /// Polling i/o error.
+    Poll(io::Error),
+    /// The specified pollable already registered.
+    AlreadyExists,
+    /// The specified pollable is not registered.
+    NotFound(Pollable),
+    /// Error while writing the api channel eventfd.
+    ApiChannelFd(io::Error),
+    /// Error while sending data on the api channel.
+    ApiChannelTx,
+    /// Api channel disconnected.
+    ApiChannelDisconnect,
+}
+
+impl std::fmt::Debug for Error {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        use self::Error::*;
+
+        match self {
+            EpollCreate => write!(f, "Unable to create epoll fd."),
+            Poll(err) => write!(f, "Error during epoll call: {}", err),
+            AlreadyExists => write!(f, "A handler for the specified pollable already exists"),
+            ApiChannelFd(err) => write!(f, "Error while writing Api channel event fd: {}", err),
+            ApiChannelTx => write!(f, "Error while sending on the Api channel"),
+            ApiChannelDisconnect => write!(f, "Error while reading the Api channel: disconnected"),
+            NotFound(pollable) => write!(
+                f,
+                "A handler for the specified pollable {} was not found",
+                pollable.as_raw_fd()
+            ),
+        }
+    }
+}
+
+struct EventHandlerData {
+    data: EventRegistrationData,
+    handler: WrappedHandler,
+}
+
+impl EventHandlerData {
+    fn new(data: EventRegistrationData, handler: WrappedHandler) -> EventHandlerData {
+        EventHandlerData { data, handler }
+    }
+}
+
+/// A trait to express the ability to respond to I/O event readiness
+/// using callbacks.
+pub trait EventHandler: Send {
+    /// Handle a read event (EPOLLIN).
+    fn handle_read(&mut self, _source: Pollable) -> Option<Vec<PollableOp>> {
```

Comment:
Can we return a `Vec<PollableOp>` instead of an `Option<Vec<PollableOp>>`? It would make the code neater in the sense that you don't have to check if the return type is none. For the polly usecase, it should be safe to assume that an empty vector of PollableOp is the equivalent of None.